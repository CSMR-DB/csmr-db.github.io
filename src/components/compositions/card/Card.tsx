import React, { ReactNode } from 'react'
import styled, {
  StyledComponent,
  DefaultTheme,
  css,
  FlattenSimpleInterpolation,
} from 'styled-components'
import { Tween } from 'react-gsap'

export interface ICardProps {
  theme?: DefaultTheme
  children: ReactNode
  index?: number
}

export const StyledCard: StyledComponent<
  'article',
  any,
  ICardProps,
  never
> = styled.article`
  position: relative;
  height: 100%;
  border-radius: 0.25rem;
  display: grid;
  grid-template-areas: 'header' 'body' 'footer';
  grid-template-rows: auto 1fr auto;
  overflow: hidden;
  box-shadow: 0 0.075rem 0.075rem rgba(0, 0, 0, 0.1),
    0 0.125rem 0.125rem rgba(0, 0, 0, 0.25);
  transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);

  & header img {
    transition: all 0.2s ease-in-out !important;
  }

  & h1 {
    font-size: 1.1rem;
    margin: 0;
  }

  & h2 {
    font-size: 1rem;
    font-style: italic;
    margin: 0;
    color: #aaa;
    margin-bottom: 1rem;
  }

  & p {
    text-align: left;
  }

  ${({ theme }: ICardProps): FlattenSimpleInterpolation => css`
    background: ${theme!.palette.light.normal};

    @media ${theme!.breakpoints.min.smartphone} {
      text-align: justify;
    }

    @media ${theme!.breakpoints.min.desktop} {
      &:hover {
        box-shadow: 0 0.125rem 0.25rem rgba(0, 0, 0, 0.25),
          0 0.25rem 0.25rem rgba(0, 0, 0, 0.2);

        & header img {
          transform: scale(1.1);
        }
      }
    }
  `}
`

export function Card({ children, index = 0 }: ICardProps): JSX.Element {
  return (
    <Tween
      from={{ scale: 0.95, opacity: 0 }}
      duration={0.5}
      delay={0.25 + index * 0.125}
      ease="back"
    >
      <StyledCard>{children}</StyledCard>
    </Tween>
  )
}
