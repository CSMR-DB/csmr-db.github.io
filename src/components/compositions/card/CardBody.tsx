import styled, { StyledComponent } from 'styled-components'

export const CardBody: StyledComponent<'main', any, {}, never> = styled.main`
  position: relative;
  padding: 2em;
  grid-area: 'body';
`
