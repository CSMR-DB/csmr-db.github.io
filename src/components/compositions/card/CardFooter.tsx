import styled, {
  StyledComponent,
  FlattenSimpleInterpolation,
  css,
} from 'styled-components'

interface ICardFooterProps {
  $style?: boolean
}

export const CardFooter: StyledComponent<
  'footer',
  any,
  ICardFooterProps,
  never
> = styled.footer`
  ${({ $style: style = false }: ICardFooterProps): FlattenSimpleInterpolation =>
    style
      ? css`
          background: #eee;
          padding: 2rem;
          border-top: 1px solid #ddd;
          grid-area: footer;
        `
      : css`
          padding: 0 2rem 2rem;
        `}
`
