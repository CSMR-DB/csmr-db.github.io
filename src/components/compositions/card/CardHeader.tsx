import styled, { StyledComponent } from 'styled-components'

export const CardHeader: StyledComponent<
  'header',
  any,
  {},
  never
> = styled.header`
  grid-area: 'header';
`
