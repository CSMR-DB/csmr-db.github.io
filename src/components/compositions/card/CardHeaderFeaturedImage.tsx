import styled, {
  StyledComponent,
  css,
  FlattenSimpleInterpolation,
} from 'styled-components'
import { MaxHeightProperty } from 'csstype'

interface ICardHeaderFeaturedImageProps {
  maxHeight?: MaxHeightProperty<1>
}

export const CardHeaderFeaturedImage: StyledComponent<
  'div',
  any,
  ICardHeaderFeaturedImageProps,
  never
> = styled.div`
  overflow: hidden;

  ${({
    maxHeight = '18rem',
  }: ICardHeaderFeaturedImageProps): FlattenSimpleInterpolation => css`
    max-height: ${maxHeight};
    min-height: ${maxHeight};

    & > * {
      max-height: ${maxHeight};
      min-height: ${maxHeight};
    }
  `}
`
