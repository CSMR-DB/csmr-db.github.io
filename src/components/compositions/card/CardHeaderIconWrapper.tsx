import styled, {
  StyledComponent,
  css,
  FlattenSimpleInterpolation,
} from 'styled-components'
import { BackgroundColorProperty } from 'csstype'

interface ICardHeaderIconWrapperProps {
  $backgroundColor: BackgroundColorProperty
}

export const CardHeaderIconWrapper: StyledComponent<
  'div',
  any,
  ICardHeaderIconWrapperProps,
  never
> = styled.div`
  border-radius: 999px;
  overflow: hidden;
  width: 4em;
  padding: 0.5em;

  ${({
    $backgroundColor: backgroundColor = '#DDD',
  }: ICardHeaderIconWrapperProps): FlattenSimpleInterpolation => css`
    background: ${backgroundColor};
  `}
`
