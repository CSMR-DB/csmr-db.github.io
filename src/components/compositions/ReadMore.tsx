import React from 'react'
import styled, {
  StyledComponent,
  DefaultTheme,
  css,
  FlattenSimpleInterpolation,
} from 'styled-components'

import { Card } from './card/Card'
import { CardBody } from './card/CardBody'
import { StyledLink } from '../StyledLink'

export interface IReadMoreProps {
  theme?: DefaultTheme
  title: string
  link: {
    path: string
    text: string
  }
  $_testid?: string
}

export const StyledReadMore: StyledComponent<
  'div',
  any,
  Pick<IReadMoreProps, 'theme'>,
  never
> = styled.div`
  display: flex;

  ${({ theme }: IReadMoreProps): FlattenSimpleInterpolation => css`
    flex-direction: column;
    gap: 1rem;

    @media ${theme!.breakpoints.min.smartphone} {
      flex-direction: row;
      justify-content: space-between;
      align-items: baseline;
    }
  `}
`

export function ReadMore({
  title,
  link,
  $_testid,
}: IReadMoreProps): JSX.Element {
  return (
    <Card $_testid={$_testid}>
      <CardBody>
        <StyledReadMore>
          <h1>{title}</h1>
          <StyledLink to={link.path} $_testid={'ReadMoreLink'}>
            {link.text}
          </StyledLink>
        </StyledReadMore>
      </CardBody>
    </Card>
  )
}
