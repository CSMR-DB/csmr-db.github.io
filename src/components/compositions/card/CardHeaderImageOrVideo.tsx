import React from 'react'
import Image from 'gatsby-image'
import { MaxHeightProperty } from 'csstype'

import { FileChildImageSharp } from '../../../types/graphql.types'
import {
  AnyExoticRefComponent,
  AnyExoticRefTargets,
} from '../../../types/react.types'
import { StaticDataManager } from '../../../data/StaticDataManager'

import { CardHeaderFeaturedImage } from './CardHeaderFeaturedImage'
import { DynamicImage } from '../DynamicImage'
import { StyledYoutube } from '../../StyledYoutube'

interface IImageOrVideo {
  image?: string | FileChildImageSharp
  video?: string
  maxHeight?: MaxHeightProperty<1>
}

export const CardHeaderImageOrVideo: AnyExoticRefComponent<IImageOrVideo> = React.forwardRef(
  (
    { image, video, maxHeight }: IImageOrVideo,
    ref: AnyExoticRefTargets
  ): JSX.Element => (
    <CardHeaderFeaturedImage $maxHeight={maxHeight} ref={ref}>
      {image && typeof image === 'string' ? (
        <DynamicImage
          path={image}
          dynamicImages={StaticDataManager.dynamicImages}
        />
      ) : (
        (image && typeof image !== 'string' && (
          <Image fluid={image.childImageSharp.fluid} />
        )) ||
        (video && (
          <StyledYoutube
            opts={{
              width: '960',
              height: '290',
            }}
            videoId={video.replace('https://www.youtube.com/embed/', '')}
          ></StyledYoutube>
        ))
      )}
    </CardHeaderFeaturedImage>
  )
)
