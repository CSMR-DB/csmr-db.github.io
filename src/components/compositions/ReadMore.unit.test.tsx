import React, { Component } from 'react'
import { ThemeProvider } from 'styled-components'
import { create, ReactTestRenderer } from 'react-test-renderer'
import { mount, ReactWrapper } from 'enzyme'

import { StaticDataManager } from '../../data/StaticDataManager'

import { ReadMore } from './ReadMore'
import { StyledLink } from '../StyledLink'
import { Card } from './card/Card'

jest.mock('react-gsap')

describe('ReadMore', (): void => {
  it('should render a read more Card with link', (): void => {
    const comp: ReactTestRenderer = create(
      <ThemeProvider theme={StaticDataManager.theme}>
        <ReadMore title={'Hello'} link={{ path: '/hello', text: 'World' }} />
      </ThemeProvider>
    )

    expect(comp).toMatchSnapshot()
  })

  it('should render internally', (): void => {
    const comp: ReactWrapper<any, Readonly<{}>, Component<{}, {}, any>> = mount(
      <ThemeProvider theme={StaticDataManager.theme}>
        <ReadMore title={'Hello'} link={{ path: '/hello', text: 'World' }} />
      </ThemeProvider>
    )

    expect(comp.find(Card).length).toEqual(1)
    expect(comp.find(Card).find('h1').length).toEqual(1)
    expect(comp.find(Card).find('h1').text()).toEqual('Hello')
    expect(comp.find(Card).find(StyledLink).length).toEqual(1)
    expect(comp.find(Card).find(StyledLink).text()).toEqual('World')
  })
})
