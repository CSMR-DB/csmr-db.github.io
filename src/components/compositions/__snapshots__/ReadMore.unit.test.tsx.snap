// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadMore should render a read more Card with link 1`] = `
.c0 {
  position: relative;
  height: 100%;
  border-radius: 0.25rem;
  display: grid;
  grid-template-areas: 'header' 'body' 'footer';
  grid-template-rows: auto 1fr auto;
  overflow: hidden;
  box-shadow: 0 0.075rem 0.075rem rgba(0,0,0,0.1), 0 0.125rem 0.125rem rgba(0,0,0,0.25);
  -webkit-transition: all 0.3s cubic-bezier(0.25,0.8,0.25,1);
  transition: all 0.3s cubic-bezier(0.25,0.8,0.25,1);
  background: #fff;
}

.c0 header img {
  -webkit-transition: all 0.2s ease-in-out !important;
  transition: all 0.2s ease-in-out !important;
}

.c0 h1 {
  font-size: 1.1rem;
  margin: 0;
}

.c0 h2 {
  font-size: 1rem;
  font-style: italic;
  margin: 0;
  color: #aaa;
  margin-bottom: 1rem;
}

.c0 p {
  text-align: left;
}

.c1 {
  grid-area: 'body';
  position: relative;
  padding: 2rem;
}

.c3 {
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
  color: #f00 !important;
}

.c3:hover {
  color: #d00;
}

.c3.active {
  color: #b00;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1rem;
}

@media only screen and (min-width:600px) {
  .c0 {
    text-align: justify;
  }
}

@media only screen and (min-width:1200px) {
  .c0:hover {
    box-shadow: 0 0.125rem 0.25rem rgba(0,0,0,0.25), 0 0.25rem 0.25rem rgba(0,0,0,0.2);
  }

  .c0:hover header img {
    -webkit-transform: scale(1.1);
    -ms-transform: scale(1.1);
    transform: scale(1.1);
  }
}

@media only screen and (min-width:600px) {
  .c2 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
    -webkit-align-items: baseline;
    -webkit-box-align: baseline;
    -ms-flex-align: baseline;
    align-items: baseline;
  }
}

<article
  className="c0"
>
  <main
    className="c1"
  >
    <div
      className="c2"
    >
      <h1>
        Hello
      </h1>
      <a
        className="c3"
        href="/hello"
      >
        World
      </a>
    </div>
  </main>
</article>
`;
